src/core/maprenderer/qgsmaprendererjob.cpp

```diff
@@ -651,66 +680,124 @@ std::vector< LayerRenderJob > QgsMapRendererJob::prepareSecondPassJobs( std::vec
+  // Prepare label mask images
+  for ( int maskId = 0; maskId < labelJob.maskIdProvider.size(); maskId++ )
   {
-    if ( job.img == nullptr )
+    if ( forceVector && !labelHasEffects[ maskId ] )
     {
-      job.context()->setPainter( allocateImageAndPainter( job.layerId, job.img ) );
+      // set a painter to get all masking instruction in order to later clip masked symbol layer
+      QPainter *painter = new QPainter( new QgsMaskPaintDevice( true ) );
+      labelJob.context.setMaskPainter( painter, maskId );
     }
-    if ( layerHasMask.contains( job.layerId ) )
+    else
     {
       // Note: we only need an alpha channel here, rather than a full RGBA image
-      job.context()->setMaskPainter( allocateImageAndPainter( job.layerId, job.maskImage ) );
-      job.maskImage->fill( 0 );
+      QImage *maskImage;
+      labelJob.context.setMaskPainter( allocateImageAndPainter( QStringLiteral( "label mask" ), maskImage ), maskId );
+      maskImage->fill( 0 );
+      labelJob.maskImages.push_back( maskImage );
     }
   }
+  labelJob.context.setMaskIdProvider( &labelJob.maskIdProvider );
+
+  // Prepare second pass jobs
+  // - For raster rendering or vector rendering if effects are involved
+  // 1st pass, 2nd pass and mask are rendered in QImage and composed in composeSecondPass
+  // - For vector rendering if no effects are involved
+  // 1st pass is rendered in QImage, clip paths are generated according to mask and used during
+  // masked symbol layer rendering during second pass, which is rendered in QPicture, second
+  // pass job imgPic
 
   // Allocate an image for labels
-  if ( labelJob.img == nullptr )
+  if ( labelJob.img == nullptr && !forceVector )
   {
     labelJob.img = allocateImage( QStringLiteral( "labels" ) );
   }
+  else if ( labelJob.imgPic == nullptr && forceVector )
+  {
+    labelJob.imgPic.reset( new QPicture() );
+  }
 
-  // Prepare label mask images
-  for ( int maskId = 0; maskId < labelJob.maskIdProvider.size(); maskId++ )
+  // first we initialize painter and mask painter for all jobs
+  for ( LayerRenderJob &job : firstPassJobs )
   {
-    QImage *maskImage;
-    labelJob.context.setMaskPainter( allocateImageAndPainter( QStringLiteral( "label mask" ), maskImage ), maskId );
-    maskImage->fill( 0 );
-    labelJob.maskImages.push_back( maskImage );
+    job.needRasterization = false;
+
+    auto it = maskedSymbolLayers.find( job.layerId );
+    if ( it != maskedSymbolLayers.end() )
+    {
+      QList<MaskSource> &sourceList = it->second;
+      for ( MaskSource &source : sourceList )
+      {
+        job.needRasterization |= source.hasEffects;
+      }
+    }
+
+    // update first pass job painter and device if needed
+    const bool isRasterRendering = !forceVector || job.needRasterization || ( job.renderer && job.renderer->forceRasterRender() );
+    if ( isRasterRendering && !job.img )
+    {
+      job.context()->setPainter( allocateImageAndPainter( job.layerId, job.img ) );
```

Comment:
Job is not always the ownership of the painter & device it is working with (contrary to mask painter/device for which I made the modifications you suggest) because a painter can be passed in *prepareJobs*. And so we should only store a painter when we need an other one (blending, masking, parralel rendering). 

The actual logic was to delete the painter when an image was created, so I follow the same logic with the qpicture, and I would prefer to keep it that way and not rework all the existing logic.